### PR TITLE
Low Level Optimisations

### PR DESCRIPTION
**Pull request by jmamma**
Going to submit low level optimisations here @yatli 

Anything to speed up the ISR routines.

First one is:

There are only two memory banks, so we don't care what the absolute
value is.

### CODE REVIEW

File: avr/cores/megacommand/MD/MD.cpp [hunk 1]
```diff
@@ -162,10 +162,10 @@ void MDClass::triggerTrack(uint8_t track, uint8_t velocity) {
...
 void MDClass::setTrackParam(uint8_t track, uint8_t param, uint8_t value) {
-  if (global.baseChannel > 15)
-    return;
-  if ((track > 15) || (param > 33))
+/*  if (global.baseChannel > 15)
     return;
+    if ((track > 15) || (param > 33))
+    return;*/
 
   uint8_t channel = track >> 2;
```
Comment (yatli) [inline avr/cores/megacommand/MD/MD.cpp:170]: track / 4? now that we know the trick ;)

Comment (jmamma) [reply avr/cores/megacommand/MD/MD.cpp:170]: Did I miss something important. I remember you said that 
n / 4 didn't optimize to n >> 2

If you can see optimisation improvements for setTrackParam please assist :smiley:

**Comment by yatli** [comment]
I propose to split RingBuffer into two classes, one for built-in buffer and one for external ptr.

**Comment by jmamma** [comment]
> I propose to split RingBuffer into two classes, one for built-in buffer and one for external ptr.

Yeah that's probably a better approach

File: avr/cores/megacommand/MD/MD.cpp (hunk 1 above)
Comment (yatli) [reply avr/cores/megacommand/MD/MD.cpp:170]: It's not "n/4 didn't optimize to n>>2", rather the opposite:
n/4 will perform better than n>>2 (!)

Comment (yatli) [reply avr/cores/megacommand/MD/MD.cpp:170]: >improvements for setTrackParam

sure, will go through that

Comment (jmamma) [reply avr/cores/megacommand/MD/MD.cpp:170]: @yatli I worked out what's happening

Both n /4 and n >> 2 will be optimized to the same instruction set when the
-Os compiler flag is enabled

https://godbolt.org/z/sRIPhS

Comment (jmamma) [reply avr/cores/megacommand/MD/MD.cpp:170]: we are already compiling with -Os

Comment (jmamma) [reply avr/cores/megacommand/MD/MD.cpp:170]: This one is good read. https://ww1.microchip.com/downloads/en/AppNotes/doc8453.pdf

Comment (yatli) [reply avr/cores/megacommand/MD/MD.cpp:170]: check this out: https://godbolt.org/z/luMJFr

Comment (yatli) [reply avr/cores/megacommand/MD/MD.cpp:170]: When the shifted value is used as addressing index (which must be 16bit), shift generates two more instruction than the division version.

(I've no idea why this is happening... But let's just sacrifice the young goat, for fundamentally technical reasons. :D)

Comment (jmamma) [reply avr/cores/megacommand/MD/MD.cpp:170]: 😲

Comment (yatli) [reply avr/cores/megacommand/MD/MD.cpp:170]: I just took another look:

```
1:      asr r31
        ror r30
        dec r18
        brne 1b
```

so it's not just two more instructions. it's a loop

😂

**Comment by jmamma** [comment]
i've merged this branch in to dev

**Pull request closed without merging**